core/shared/src/main/scala/zio/FiberRefs.scala

```diff
@@ -63,18 +64,22 @@ final class FiberRefs private (
    */
   def forkAs(childId: FiberId.Runtime): FiberRefs =
     if (needsTransformWhenForked) {
-      var modified = false
-      val childMap = fiberRefLocals.transform { case (fiberRef, entry @ Value(stack, depth)) =>
-        val oldValue = stack.head.value.asInstanceOf[fiberRef.Value]
-        val newValue = fiberRef.patch(fiberRef.fork)(oldValue)
-        if (oldValue == newValue) entry
-        else {
-          modified = true
-          Value(::(StackEntry(childId, newValue, 0), stack), depth + 1)
+      val childMap = fiberRefLocals.transform { (fiberRef, entry) =>
+        val fork = fiberRef.fork
+        if (fork.asInstanceOf[AnyRef] eq ZIO.identityFn[Any]) {
+          entry
+        } else {
+          import entry.{depth, stack}
+
+          type T = fiberRef.Value with AnyRef
+          val oldValue = stack.head.value.asInstanceOf[T]
+          val newValue = fiberRef.patch(fork)(oldValue).asInstanceOf[T]
```

Comment:
minor: Can't you just cast `newValue` in the case that it's used? Or are these not the same type prior to casting?

Reply:
We have to cast it otherwise `eq` doesn't work (since the generic type is not a subtype of `AnyRef`)